src/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -0,0 +1,98 @@
+module Database.LSMTree.Internal.Chunk
+(
+    -- * Chunks
+    Chunk (Chunk),
+
+    -- * Balers
+    Baler,
+    createBaler,
+    feedBaler,
+    readBalerRemnant
+)
+where
+
+import           Prelude hiding (concat, length, splitAt)
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.STRef.Strict (STRef, newSTRef, readSTRef, writeSTRef)
+import           Data.Vector.Primitive (Vector, concat, length, slice, splitAt)
+import           Data.Word (Word8)
+
+-- * Chunks
+
+-- | A chunk of bytes, typically output during incremental index serialisation.
+newtype Chunk = Chunk (Vector Word8)
+
+-- * Balers
+
+-- Mutable state of a baler.
+data BalerState = BalerState
+                      ![Vector Word8] -- Reverse list of queued blocks
+                      !Int            -- Total size of queued blocks
+
+{-|
+    An object that receives blocks of bytes and repackages them into chunks of
+    preferably a given maximum size.
+-}
+data Baler s = Baler
+                   !Int                  -- Maximum chunk size
+                   !(STRef s BalerState) -- Reference to the mutable state
+
+-- | Creates a new baler.
+createBaler :: Int            -- ^ Maximum chunk size
+            -> ST s (Baler s) -- ^ Creation of the baler
+createBaler maxChunkSize = Baler maxChunkSize <$> newSTRef (BalerState [] 0)
+
+{-|
+    Feeds a baler a block of bytes.
+
+    Bytes received by a bailer are generally queued for later output, but if
+    feeding new bytes makes the queued content exceed the maximum chunk size
+    then as many chunks of maximum size as possible are output.
+-}
+feedBaler :: Vector Word8 -> Baler s -> ST s [Chunk]
+feedBaler block (Baler maxChunkSize stateRef) = do
+    BalerState remnantBlocksRev remnantSize <- readSTRef stateRef
+    let
+
+        totalSize :: Int
+        totalSize = remnantSize + length block
+
+    if totalSize < maxChunkSize
+        then do
+                 writeSTRef stateRef
+                            (BalerState (block : remnantBlocksRev) totalSize)
+                 return []
+        else do
+                 let
+
+                     newRemnantSize :: Int
+                     newRemnantSize = totalSize `mod` maxChunkSize
```

Comment:
```suggestion
                     !newRemnantSize = totalSize `mod` maxChunkSize
```

Reply:
I was inclined to write the same as I wrote [above](https://github.com/IntersectMBO/lsm-tree/pull/296#discussion_r1680779780) regarding `totalSize`, but then I figured that it is much harder to convince yourself that `newRemnantSize` will be evaluated promptly (in fact, it might not even be evaluated promptly without adding the `$!` you suggested to add [above](https://github.com/IntersectMBO/lsm-tree/pull/296#discussion_r1680797032)). Maybe it is indeed better to be more generous with `!`s and make intended strictness more explicit.

However, shouldn’t then some other patterns get `!`s added as well, for example `output`? What is actually the general policy regarding what should be strict and what not? I guess that things stored in mutable variables should usually contain no thunks, as such thunks would stay beyond the execution of the `ST` (or `IO`) action that created them with no chance for outside code to force their evaluation. But what about things like the list of chunks that `feedBaler` creates? Should this be lazy? I’m inclined to say “yes”, but I guess that it would be a good idea to make sure that the underlying `output` vector has been evaluated by the time the execution of `feedBaler` ends. What is your take on these issues, and is there a general concept that is followed throughout the `lsm-tree` package?

Reply:
This all sounds very sensible to me. 🙂

Reply:
I’ve made the suggested change in 9c8dec880f539cc47027a16e5872696bc4240850.